ApiIndex/ApiModel.cs

```diff
@@ -0,0 +1,218 @@
+            Id = directory.Replace('/', '.');
+
+            Title = serviceConfig.Title;
+            Description = serviceConfig.Documentation?.Summary?.Replace('\n', ' ') ?? "<UNKNOWN - NO SERVICE CONFIG DOCUMENTATION SUMMARY>";
+            Version = directory.Split('/').Last();
+            MajorVersion = ApiMajorVersionPattern.Match(Version).Value;
+            HostName = serviceConfig.Name;
+
+            Services = Files.SelectMany(file => file.Services)
+                .OrderBy(service => service.FullName)
+                .Select(service => new ServiceModel(service))
+                .ToList()
+                .AsReadOnly();
+        }
+
+        public static IReadOnlyList<ApiModel> Load(FileDescriptorSet descriptorSet, string googleApisRoot)
+        {
+            var descriptorByteStrings = descriptorSet.File.Select(proto => proto.ToByteString());
+            var allDescriptors = FileDescriptor.BuildFromByteStrings(descriptorByteStrings).ToList().AsReadOnly();
+
+            var models = new List<ApiModel>();
+            AddServiceConfigsRecursively(googleApisRoot);
+
+            return models.OrderBy(model => model.Id, StringComparer.Ordinal).ToList().AsReadOnly();
+
+            void AddServiceConfigsRecursively(string directory)
+            {
+                // Only load service config files from versioned directories (e.g. "google/spanner/v1")
+                if (ApiMajorVersionPattern.IsMatch(Path.GetFileName(directory)))
+                {
+                    string relativeDirectory = Path.GetRelativePath(googleApisRoot, directory).Replace('\\', '/');
+                    var configsInDirectory = System.IO.Directory.GetFiles(directory, "*.yaml")
+                        .Select(ServiceConfig.TryLoadFile)
+                        .Where(config => config != null)
+                        .ToList();
+                    switch (configsInDirectory.Count)
+                    {
+                        case 0: break;
+                        case 1:
+                            models.Add(new ApiModel(relativeDirectory, configsInDirectory[0], allDescriptors));
+                            break;
+                        default:
+                            models.Add(new ApiModel(relativeDirectory, configsInDirectory[0], allDescriptors));
+                            Console.WriteLine($"Directory '{relativeDirectory}' contains {configsInDirectory.Count} service (API) config YAML files.");
+                            break;
+                    }
+                }
+                else
+                {
+                    // Note: we assume there are no "nested" APIs
+                    foreach (var subdir in System.IO.Directory.GetDirectories(directory))
+                    {
+                        AddServiceConfigsRecursively(subdir);
+                    }
+                }
+            }
+        }
+
+        public V1.Api ToV1Api() =>
```

Comment:
Might be overkill, but can this be moved to ApiIndex.V1 as an extension method or by creating a new class there that inherits from `ApiIndex.ApiModel`.
If anything the name `ToV1Api` is very confusing, at least to me, I read the version as if it were the API version and not the ApiIndex version.
Also, if different ApiIndex versions get to coexist, does that mean that there'll be several `ToV*Api` methods in this class?
And as it stands, if we want to get rid of ApiIndex.V1 (in favour of ApiIndex.V2 for instance), this would still have to take a breaking change to remove this public method, so the apparent independece of `Google.Cloud.Tools.ApiIndex` from the index protos, is really not there if using this pattern for the conversion methods.

Reply:
We can't make it an extension method, but we could create a partial method of `FromModel` in `ApiIndex.V1.Api` potentially. I tend to find "To" conversions more discoverable than "From" conversions though. How about `ToProtoV1` as a name to use everywhere?

> Also, if different ApiIndex versions get to coexist, does that mean that there'll be several ToV*Api methods in this class?

Yes.

> And as it stands, if we want to get rid of ApiIndex.V1 (in favour of ApiIndex.V2 for instance), this would still have to take a breaking change to remove this public method

This isn't a library, so "breaking change" isn't really a problem.

The intention is that these model classes should always contain everything required to generate "all current schema versions", and have that information in the most convenient form. That just makes it easier to support multiple versions without too much duplication.